lighthouse-core/fraggle-rock/gather/navigation-runner.js

```diff
@@ -5,116 +5,220 @@
+  } else {
+    await collectPhaseArtifacts({phase: 'getArtifact', ...phaseState});
+
+    const artifacts = await awaitArtifacts(phaseState.artifactState);
+    return {artifacts, warnings, pageLoadError: undefined};
+  }
+}
 
 /**
  * @param {NavigationContext} navigationContext
  */
 async function _navigation(navigationContext) {
   const artifactState = getEmptyArtifactState();
-  const options = {
+  const phaseState = {
     gatherMode: /** @type {'navigation'} */ ('navigation'),
     driver: navigationContext.driver,
+    computedCache: navigationContext.computedCache,
     artifactDefinitions: navigationContext.navigation.artifacts,
     artifactState,
   };
 
-  await _setupNavigation(navigationContext);
-  await collectPhaseArtifacts({phase: 'startInstrumentation', ...options});
-  await collectPhaseArtifacts({phase: 'startSensitiveInstrumentation', ...options});
-  await _navigate(navigationContext);
-  await collectPhaseArtifacts({phase: 'stopSensitiveInstrumentation', ...options});
-  await collectPhaseArtifacts({phase: 'stopInstrumentation', ...options});
-  await collectPhaseArtifacts({phase: 'getArtifact', ...options});
+  const setupResult = await _setupNavigation(navigationContext);
+  await collectPhaseArtifacts({phase: 'startInstrumentation', ...phaseState});
+  await collectPhaseArtifacts({phase: 'startSensitiveInstrumentation', ...phaseState});
+  const navigateResult = await _navigate(navigationContext);
+  await collectPhaseArtifacts({phase: 'stopSensitiveInstrumentation', ...phaseState});
+  await collectPhaseArtifacts({phase: 'stopInstrumentation', ...phaseState});
+  await _cleanupNavigation(navigationContext);
 
-  const artifacts = await awaitArtifacts(artifactState);
-  return {artifacts};
+  return _computeNavigationResult(navigationContext, phaseState, setupResult, navigateResult);
 }
 
 /**
- * @param {{driver: Driver, config: LH.Config.FRConfig, requestedUrl: string}} args
+ * @param {{driver: Driver, config: LH.Config.FRConfig, requestedUrl: string; computedCache: NavigationContext['computedCache']}} args
  */
-async function _navigations({driver, config, requestedUrl}) {
+async function _navigations({driver, config, requestedUrl, computedCache}) {
   if (!config.navigations) throw new Error('No navigations configured');
 
-  /** @type {Partial<LH.GathererArtifacts>} */
+  /** @type {Partial<LH.FRArtifacts & LH.FRBaseArtifacts>} */
   const artifacts = {};
 
   for (const navigation of config.navigations) {
     const navigationContext = {
       driver,
       navigation,
       requestedUrl,
+      config,
+      computedCache,
     };
 
     const navigationResult = await _navigation(navigationContext);
+    if (navigationResult.pageLoadError && navigation.loadFailureMode === 'fatal') {
+      artifacts.PageLoadError = navigationResult.pageLoadError;
+      break;
+    }
+
+    // TODO(FR-COMPAT): merge RunWarnings between navigations
```

Comment:
What's preventing us from doing this now?

Reply:
PR size, the handling of runwarnings has a small cascade of other changes that already fit in well with the base artifacts cleanup I have left and was planning on putting up today.